src/Build.UnitTests/BackEnd/CacheAggregator_Tests.cs

```diff
@@ -0,0 +1,315 @@
+        public void RejectCachesWithMismatchedIds()
+        {
+            // one entry in each cache but different config ids
+
+            var configCache = new ConfigCache();
+            configCache.AddConfiguration(new BuildRequestConfiguration(configId: 1, new BuildRequestData("path", new Dictionary<string, string>(){["p"] = "v"}, "13", new []{"a", "b"}, null), "13"));
+
+            var resultsCache = new ResultsCache();
+            var buildResult = new BuildResult(new BuildRequest(1, 2, configurationId: 2, new List<string>(){"a", "b"}, null, BuildEventContext.Invalid, null));
+            buildResult.AddResultsForTarget("a", GetNonEmptySucceedingTargetResult());
+            resultsCache.AddResult(buildResult);
+
+            aggregator.Add(configCache, resultsCache);
+
+            using (var env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDONOTLAUNCHDEBUGGER", "1");
+                var e = Should.Throw<InternalErrorException>(() => aggregator.Aggregate());
+                e.Message.ShouldContain("Each result should have a corresponding configuration. Otherwise the caches are not consistent");
+            }
+        }
+
+                [Fact]
+        public void RejectCollidingConfigurationsFromSeparateCaches()
+        {
+            var configCache1 = new ConfigCache();
+            configCache1.AddConfiguration(new BuildRequestConfiguration(configId: 1, new BuildRequestData("path", new Dictionary<string, string>(){["p"] = "v"}, "13", new []{"a", "b"}, null), "13"));
+
+            var resultsCache1 = new ResultsCache();
+            var buildResult11 = new BuildResult(new BuildRequest(1, 2, configurationId: 1, new List<string>(){"a", "b"}, null, BuildEventContext.Invalid, null));
+            buildResult11.AddResultsForTarget("a", GetNonEmptySucceedingTargetResult());
+            resultsCache1.AddResult(buildResult11);
+
+            var configCache2 = new ConfigCache();
+            configCache2.AddConfiguration(new BuildRequestConfiguration(configId: 1, new BuildRequestData("path", new Dictionary<string, string>(){["p"] = "v"}, "13", new []{"a", "b"}, null), "13"));
+
+            var resultsCache2 = new ResultsCache();
+            var buildResult21 = new BuildResult(new BuildRequest(1, 2, configurationId: 1, new List<string>(){"e", "f"}, null, BuildEventContext.Invalid, null));
+            buildResult21.AddResultsForTarget("a", GetNonEmptySucceedingTargetResult());
+            resultsCache2.AddResult(buildResult21);
+
+            aggregator.Add(configCache1, resultsCache1);
+            aggregator.Add(configCache2, resultsCache2);
+
+            using (var env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDONOTLAUNCHDEBUGGER", "1");
+                var e = Should.Throw<ArgumentException>(() => aggregator.Aggregate());
+                e.Message.ShouldContain("An item with the same key has already been added");
```

Comment:
You're right, I am relying on an error spewed out by a dictionary inside the ConfigCache. The aggregator should explicitly call this out. It would make aggregating more expensive (an extra path+globalProp+toolsversion match for each inserted config against all the entries in the aggregated cache), but probably worth it (until telemetry tells us it's slow, in which case we might as well move the assert to debug only).

In practice this would appear if the higher level engine sends two caches containing the same config identity. Also hard to tell if this should be an msbuild assertion, or a user error.

Reply:
Made this an msbuild error wand will get a better idea after we see some real world failures.